File: packages/KNIRVBASE/ts/src/components/collection/distributed_collection.ts
import { VectorClock, increment, merge } from '../clock/vector_clock';
import { Network } from '../network/network_manager';
import { Storage } from '../storage/storage';
import {
  DistributedDocument,
  CRDTOperation,
  OperationType,
  SyncState,
  ProtocolMessage,
  MessageType,
  EntryType
} from '../types/types';
import { ToDistributed, ToRegular, ApplyOperation } from '../resolver/crdt_resolver';
import { ModalityType } from '../storage/nrv/spec';

export interface Frame {
  id: string;
  vector: Float32Array;
  seed: Uint8Array;
  thermo: {
    tempCelsius: number;
    voltageV: number;
    freqMHz: number;
    fanRPM: number;
  };
  proof: Uint8Array;
}

// A minimal in-memory local collection implementation to keep the example self-contained
export class LocalCollection {
  private name: string;
  private store: Storage;

  constructor(name: string, store: Storage) {
    this.name = name;
    this.store = store;
  }

  async insert(doc: Record<string, any>): Promise<Record<string, any>> {
    const cloned = this.cloneMap(doc);
    await this.store.insert(this.name, cloned);
    return this.cloneMap(cloned);
  }

  async update(id: string, update: Record<string, any>): Promise<number> {
    return await this.store.update(this.name, id, update);
  }

  async delete(id: string): Promise<number> {
    return await this.store.delete(this.name, id);
  }

  async find(id: string): Promise<Record<string, any> | null> {
    return await this.store.find(this.name, id);
  }

  async findAll(): Promise<Record<string, any>[]> {
    return await this.store.findAll(this.name);
  }

  getStore(): Storage {
    return this.store;
  }

  private cloneMap(m: Record<string, any>): Record<string, any> {
    const out: Record<string, any> = {};
    for (const k in m) {
      const v = m[k];
      if (typeof v === 'object' && v !== null && !Array.isArray(v)) {
        out[k] = this.cloneMap(v);
      } else if (Array.isArray(v)) {
        out[k] = this.cloneSlice(v);
      } else {
        out[k] = v;
      }
    }
    return out;
  }

  private cloneSlice(s: any[]): any[] {
    const out: any[] = [];
    for (const e of s) {
      if (typeof e === 'object' && e !== null && !Array.isArray(e)) {
        out.push(this.cloneMap(e));
      } else if (Array.isArray(e)) {
        out.push(this.cloneSlice(e));
      } else {
        out.push(e);
      }
    }
    return out;
  }
}

// DistributedCollection manages local storage plus network synchronization
export class DistributedCollection {
  public name: string;
  private network: Network;
  private networkID: string = '';
  private syncStates: Map<string, SyncState> = new Map();
  private operationLog: CRDTOperation[] = [];
  private maxLogSize: number = 10000;
  private local: LocalCollection;

  constructor(name: string, network: Network, store: Storage) {
    this.name = name;
    this.network = network;
    this.local = new LocalCollection(name, store);
    this.setupMessageHandlers();
  }

  private setupMessageHandlers(): void {
    this.network.onMessage(MessageType.Operation, (msg: ProtocolMessage) => {
      const payload = msg.payload as Record<string, any>;
      if (!payload) return;
      const coll = payload.collection as string;
      if (coll !== this.name) return;
      const opMap = payload.operation as Record<string, any>;
      const op: CRDTOperation = opMap as any;
      this.handleRemoteOperation(op);
    });

    this.network.onMessage(MessageType.SyncRequest, (msg: ProtocolMessage) => {
      const payload = msg.payload as Record<string, any>;
      if (!payload) return;
      const coll = payload.collection as string;
      if (coll !== this.name) return;
      this.handleSyncRequest(msg);
    });

    this.network.onMessage(MessageType.SyncResponse, (msg: ProtocolMessage) => {
      const payload = msg.payload as Record<string, any>;
      if (!payload) return;
      const coll = payload.collection as string;
      if (coll !== this.name) return;
      this.handleSyncResponse(msg);
    });
  }

  async attachToNetwork(networkID: string): Promise<void> {
    if (this.networkID !== '') {
      throw new Error(`collection ${this.name} already attached to ${this.networkID}`);
    }

    await this.network.addCollectionToNetwork(networkID, this.name);

    this.networkID = networkID;
    this.syncStates.set(networkID, {
      collection: this.name,
      networkId: networkID,
      localVector: {},
      lastSync: new Date(),
      pendingOperations: [],
      stagedEntries: [],
      syncInProgress: false
    });

    await this.requestSync();
  }

  async detachFromNetwork(): Promise<void> {
    if (this.networkID === '') return;
    await this.network.removeCollectionFromNetwork(this.networkID, this.name);
    this.syncStates.delete(this.networkID);
    this.networkID = '';
  }

  async insert(doc: Record<string, any>): Promise<Record<string, any>> {
    const id = doc.id as string;
    if (!id) {
      throw new Error("document must contain 'id'");
    }
    const entryType = doc.entryType as EntryType;

    if (entryType === EntryType.Memory) {
      const payload = doc.payload as Record<string, any>;
      if (payload && 'blob' in payload) {
        // Blob will be saved by storage.insert
      }
    }

    const inserted = await this.local.insert(doc);

    if (this.networkID !== '') {
      const opPayload = ToDistributed(inserted, this.network.getPeerID());
      opPayload.entryType = entryType;

      const op: CRDTOperation = {
        id: `${this.network.getPeerID()}-${Date.now()}-${Math.random()}`,
        type: OperationType.Insert,
        collection: this.name,
        documentId: id,
        data: opPayload,
        vector: this.getCurrentVector(),
        timestamp: Date.now(),
        peerId: this.network.getPeerID(),
      };
      this.broadcastOperation(op);
    }

    return inserted;
  }

  async update(id: string, update: Record<string, any>): Promise<number> {
    const affected = await this.local.update(id, update);
    if (this.networkID !== '' && affected > 0) {
      const doc = await this.local.find(id);
      const op: CRDTOperation = {
        id: `${this.network.getPeerID()}-${Date.now()}`,
        type: OperationType.Update,
        collection: this.name,
        documentId: id,
        data: ToDistributed(doc!, this.network.getPeerID()),
        vector: this.getCurrentVector(),
        timestamp: Date.now(),
        peerId: this.network.getPeerID()
      };
      this.broadcastOperation(op);
    }
    return affected;
  }

  async delete(id: string): Promise<number> {
    const affected = await this.local.delete(id);
    if (this.networkID !== '' && affected > 0) {
      const op: CRDTOperation = {
        id: `${this.network.getPeerID()}-${Date.now()}`,
        type: OperationType.Delete,
        collection: this.name,
        documentId: id,
        data: { id, _deleted: true } as any,
        vector: this.getCurrentVector(),
        timestamp: Date.now(),
        peerId: this.network.getPeerID()
      };
      this.broadcastOperation(op);
    }
    return affected;
  }

  async find(id: string): Promise<Record<string, any> | null> {
    return await this.local.find(id);
  }

  async findAll(): Promise<Record<string, any>[]> {
    return await this.local.findAll();
  }

  getSyncState(): SyncState | null {
    if (this.networkID === '') return null;
    return this.syncStates.get(this.networkID) || null;
  }

  async forceSync(): Promise<void> {
    await this.requestSync();
  }

  streamFrames(modality?: ModalityType): AsyncGenerator<Frame, void, unknown> {
    const store = this.local.getStore() as any;
    if (store.streamFrames) {
      return store.streamFrames(this.name, modality);
    } else {
      throw new Error(`collection "${this.name}": storage backend does not support NRV streaming`);
    }
  }

  getStore(): Storage {
    return this.local.getStore();
  }

  getOperationLog(): CRDTOperation[] {
    return [...this.operationLog];
  }

  private broadcastOperation(op: CRDTOperation): void {
    if (this.networkID === '') return;

    this.operationLog.push(op);
    this.pruneOperationLog();
    const syncState = this.syncStates.get(this.networkID)!;
    syncState.localVector = increment(syncState.localVector, this.network.getPeerID());

    this.network.broadcastMessage(this.networkID, {
      type: MessageType.Operation,
      networkId: this.networkID,
      senderId: this.network.getPeerID(),
      timestamp: Date.now(),
      payload: { collection: this.name, operation: op }
    });
  }

  private async handleRemoteOperation(op: CRDTOperation): Promise<void> {
    const existing = await this.local.find(op.documentId);
    let existingDist: DistributedDocument | null = null;
    if (existing) {
      existingDist = ToDistributed(existing, op.peerId);
    }

    const result = ApplyOperation(existingDist, op);

    if (result === null) {
      await this.local.delete(op.documentId);
    } else if (result._deleted) {
      await this.local.delete(op.documentId);
    } else {
      const regular = ToRegular(result);
      if (regular) {
        await this.local.insert(regular);
      }
    }

    if (this.networkID !== '') {
      const syncState = this.syncStates.get(this.networkID)!;
      syncState.localVector = merge(syncState.localVector, op.vector);
    }
  }

  private async requestSync(): Promise<void> {
    if (this.networkID === '') {
      throw new Error('not attached to network');
    }
    const syncState = this.syncStates.get(this.networkID)!;
    if (syncState.syncInProgress) return;
    syncState.syncInProgress = true;

    await this.network.broadcastMessage(this.networkID, {
      type: MessageType.SyncRequest,
      networkId: this.networkID,
      senderId: this.network.getPeerID(),
      timestamp: Date.now(),
      payload: { collection: this.name, vector: syncState.localVector }
    });

    setTimeout(() => {
      syncState.syncInProgress = false;
    }, 10000);
  }

  private handleSyncRequest(msg: ProtocolMessage): void {
    const payload = msg.payload as Record<string, any>;
    const remoteVector = payload.vector as VectorClock;

    const missing: CRDTOperation[] = [];
    for (const op of this.operationLog) {
      const remoteClock = remoteVector[op.peerId] || 0;
      const opClock = op.vector[op.peerId] || 0;
      if (opClock > remoteClock) {
        missing.push(op);
      }
    }

    this.network.sendToPeer(msg.senderId, this.networkID, {
      type: MessageType.SyncResponse,
      networkId: this.networkID,
      senderId: this.network.getPeerID(),
      timestamp: Date.now(),
      payload: { collection: this.name, operations: missing, vector: this.syncStates.get(this.networkID)!.localVector }
    });
  }

  private handleSyncResponse(msg: ProtocolMessage): void {
    const payload = msg.payload as Record<string, any>;
    const opsIface = payload.operations as CRDTOperation[];

    for (const oi of opsIface) {
      this.handleRemoteOperation(oi);
    }

    if (this.networkID !== '') {
      const syncState = this.syncStates.get(this.networkID)!;
      syncState.syncInProgress = false;
      syncState.lastSync = new Date();
    }
  }

  private getCurrentVector(): VectorClock {
    if (this.networkID === '') return {};
    const s = this.syncStates.get(this.networkID);
    return s ? { ...s.localVector } : {};
  }

  private pruneOperationLog(): void {
    if (this.operationLog.length > this.maxLogSize) {
      this.operationLog = this.operationLog.slice(this.operationLog.length - this.maxLogSize);
    }
  }
}

File: packages/KNIRVBASE/go/internal/collection/distributed_collection.go
package collection

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"sync"
	"time"

	"github.com/knirvcorp/knirvbase/internal/clock"
	netpkg "github.com/knirvcorp/knirvbase/internal/network"
	"github.com/knirvcorp/knirvbase/internal/p2pconsensus"
	resolver "github.com/knirvcorp/knirvbase/internal/resolver"
	stor "github.com/knirvcorp/knirvbase/internal/storage"
	typ "github.com/knirvcorp/knirvbase/internal/types"

	nrv "github.com/knirvcorp/knirvbase/pkg/nrv"
)

// A minimal in-memory local collection implementation to keep the example self-contained
type LocalCollection struct {
	mu    sync.RWMutex
	name  string
	store stor.Storage
}

func NewLocalCollection(name string, store stor.Storage) *LocalCollection {
	return &LocalCollection{name: name, store: store}
}

func (c *LocalCollection) Insert(ctx context.Context, doc map[string]interface{}) (map[string]interface{}, error) {
	// Store a deep-cloned copy to avoid accidental shared references between
	// callers and the underlying storage implementation.
	cloned := cloneMap(doc)
	if err := c.store.Insert(ctx, c.name, cloned); err != nil {
		return nil, err
	}
	// Return another clone so the caller cannot mutate the stored value
	return cloneMap(cloned), nil
}

func (c *LocalCollection) Update(ctx context.Context, id string, update map[string]interface{}) (int, error) {
	return 1, c.store.Update(ctx, c.name, id, update)
}

func (c *LocalCollection) Delete(ctx context.Context, id string) (int, error) {
	return 1, c.store.Delete(ctx, c.name, id)
}

func (c *LocalCollection) Find(ctx context.Context, id string) (map[string]interface{}, error) {
	return c.store.Find(ctx, c.name, id)
}

func cloneMap(m map[string]interface{}) map[string]interface{} {
	if m == nil {
		return nil
	}
	out := make(map[string]interface{}, len(m))
	for k, v := range m {
		switch val := v.(type) {
		case map[string]interface{}:
			out[k] = cloneMap(val)
		case []interface{}:
			out[k] = cloneSlice(val)
		default:
			// primitives and unknown types are copied by value/reference as-is
			out[k] = val
		}
	}
	return out
}

func cloneSlice(s []interface{}) []interface{} {
	if s == nil {
		return nil
	}
	out := make([]interface{}, len(s))
	for i, e := range s {
		switch v := e.(type) {
		case map[string]interface{}:
			out[i] = cloneMap(v)
		case []interface{}:
			out[i] = cloneSlice(v)
		default:
			out[i] = v
		}
	}
	return out
}

// DistributedCollection manages local storage plus network synchronization
type DistributedCollection struct {
	Name         string
	network      netpkg.Network
	networkID    string
	syncStates   map[string]*typ.SyncState
	operationLog []typ.CRDTOperation
	maxLogSize   int

	consensus p2pconsensus.ConsensusManager

	local *LocalCollection
	mu    sync.Mutex
}

func NewDistributedCollection(name string, net netpkg.Network, store stor.Storage) *DistributedCollection {
	dc := &DistributedCollection{
		Name:         name,
		network:      net,
		syncStates:   make(map[string]*typ.SyncState),
		operationLog: make([]typ.CRDTOperation, 0, 1024),
		maxLogSize:   10000,
		local:        NewLocalCollection(name, store),
	}

	dc.setupMessageHandlers()
	return dc
}

func (dc *DistributedCollection) setupMessageHandlers() {
	dc.network.OnMessage(typ.MsgOperation, func(msg typ.ProtocolMessage) {
		// Basic payload validation
		payload, ok := msg.Payload.(map[string]interface{})
		if !ok {
			return
		}
		coll, _ := payload["collection"].(string)
		if coll != dc.Name {
			return
		}
		opMap, _ := payload["operation"].(map[string]interface{})
		// We assume op was encoded in a friendly form; in a real implementation we'd use typed marshaling
		var op typ.CRDTOperation
		// re-marshal map into op
		b, _ := jsonMarshal(opMap)
		_ = jsonUnmarshal(b, &op)
		go dc.handleRemoteOperation(op)
	})

	dc.network.OnMessage(typ.MsgSyncRequest, func(msg typ.ProtocolMessage) {
		payload, _ := msg.Payload.(map[string]interface{})
		coll, _ := payload["collection"].(string)
		if coll != dc.Name {
			return
		}
		go dc.handleSyncRequest(msg)
	})

	dc.network.OnMessage(typ.MsgSyncResponse, func(msg typ.ProtocolMessage) {
		payload, _ := msg.Payload.(map[string]interface{})
		coll, _ := payload["collection"].(string)
		if coll != dc.Name {
			return
		}
		go dc.handleSyncResponse(msg)
	})
}

func (dc *DistributedCollection) AttachToNetwork(networkID string) error {
	if dc.networkID != "" {
		return fmt.Errorf("collection %s already attached to %s", dc.Name, dc.networkID)
	}

	if err := dc.network.AddCollectionToNetwork(networkID, dc.Name); err != nil {
		return err
	}

	dc.networkID = networkID
	dc.syncStates[networkID] = &typ.SyncState{Collection: dc.Name, NetworkID: networkID, LocalVector: clock.NewVectorClock(), LastSync: time.Now(), PendingOperations: nil, SyncInProgress: false}

	// Request initial sync
	return dc.requestSync()
}

func (dc *DistributedCollection) DetachFromNetwork() error {
	if dc.networkID == "" {
		return nil
	}
	_ = dc.network.RemoveCollectionFromNetwork(dc.networkID, dc.Name)
	delete(dc.syncStates, dc.networkID)
	dc.networkID = ""
	return nil
}

func (dc *DistributedCollection) Insert(ctx context.Context, doc map[string]interface{}) (map[string]interface{}, error) {
	id, ok := doc["id"].(string)
	if !ok {
		return nil, errors.New("document must contain 'id'")
	}
	entryType, _ := doc["entryType"].(typ.EntryType) // Assuming validation already happened

	// For MEMORY entries, blob is handled by storage
	if entryType == typ.EntryTypeMemory {
		if payload, ok := doc["payload"].(map[string]interface{}); ok {
			if _, hasBlob := payload["blob"]; hasBlob {
				// Blob will be saved by storage.Insert
			}
		}
	}

	// The local collection handles persistence. The underlying adapter is responsible
	// for the actual file I/O for blobs.
	inserted, err := dc.local.Insert(ctx, doc)
	if err != nil {
		return nil, err
	}

	if dc.networkID != "" {
		opPayload := resolver.ToDistributed(inserted, dc.network.GetPeerID())
		opPayload.EntryType = entryType // Ensure EntryType is set on the distributed doc

		op := typ.CRDTOperation{
			ID:         fmt.Sprintf("%s-%d-%d", dc.network.GetPeerID(), time.Now().UnixMilli(), time.Now().UnixNano()%1000),
			Type:       typ.OpInsert,
			Collection: dc.Name,
			DocumentID: id,
			Data:       opPayload,
			Vector:     dc.getCurrentVector(),
			Timestamp:  time.Now().UnixMilli(),
			PeerID:     dc.network.GetPeerID(),
		}
		dc.broadcastOperation(op)
	}

	return inserted, nil
}

func (dc *DistributedCollection) Update(ctx context.Context, id string, update map[string]interface{}) (int, error) {
	affected, err := dc.local.Update(ctx, id, update)
	if err != nil {
		return 0, err
	}
	if dc.networkID != "" && affected > 0 {
		doc, _ := dc.local.Find(ctx, id)
		op := typ.CRDTOperation{ID: fmt.Sprintf("%s-%d", dc.network.GetPeerID(), time.Now().UnixMilli()), Type: typ.OpUpdate, Collection: dc.Name, DocumentID: id, Data: resolver.ToDistributed(doc, dc.network.GetPeerID()), Vector: dc.getCurrentVector(), Timestamp: time.Now().UnixMilli(), PeerID: dc.network.GetPeerID()}
		dc.broadcastOperation(op)
	}
	return affected, nil
}

func (dc *DistributedCollection) Delete(ctx context.Context, id string) (int, error) {
	affected, err := dc.local.Delete(ctx, id)
	if err != nil {
		return 0, err
	}
	if dc.networkID != "" && affected > 0 {
		op := typ.CRDTOperation{ID: fmt.Sprintf("%s-%d", dc.network.GetPeerID(), time.Now().UnixMilli()), Type: typ.OpDelete, Collection: dc.Name, DocumentID: id, Data: &typ.DistributedDocument{ID: id, Deleted: true}, Vector: dc.getCurrentVector(), Timestamp: time.Now().UnixMilli(), PeerID: dc.network.GetPeerID()}
		dc.broadcastOperation(op)
	}
	return affected, nil
}

func (dc *DistributedCollection) Find(ctx context.Context, id string) (map[string]interface{}, error) {
	return dc.local.Find(ctx, id)
}
func (dc *DistributedCollection) FindAll(ctx context.Context) ([]map[string]interface{}, error) {
	return dc.local.store.FindAll(ctx, dc.Name)
}

func (dc *DistributedCollection) GetSyncState() *typ.SyncState {
	if dc.networkID == "" {
		return nil
	}
	return dc.syncStates[dc.networkID]
}

func (dc *DistributedCollection) ForceSync() error { return dc.requestSync() }

// Private helpers

// SetConsensusManager attaches a p2pconsensus manager for P2P routing.
func (dc *DistributedCollection) SetConsensusManager(cm p2pconsensus.ConsensusManager) {
	dc.consensus = cm
}

func (dc *DistributedCollection) broadcastOperation(op typ.CRDTOperation) {
	if dc.networkID == "" {
		return
	}

	dc.mu.Lock()
	dc.operationLog = append(dc.operationLog, op)
	dc.pruneOperationLog()
	// increment local vector
	syncState := dc.syncStates[dc.networkID]
	syncState.LocalVector = clock.Increment(syncState.LocalVector, dc.network.GetPeerID())
	dc.mu.Unlock()

	// Route through p2pconsensus if available
	if dc.consensus != nil && dc.consensus.Enabled() {
		data, _ := json.Marshal(op)
		env := p2pconsensus.OperationEnvelope{
			Collection:  dc.Name,
			NetworkID:   dc.networkID,
			DocumentID:  op.DocumentID,
			Data:        data,
			VectorClock: op.Vector,
			Timestamp:   time.Now().UnixMilli(),
			PeerID:      dc.network.GetPeerID(),
		}
		_ = dc.consensus.PublishOperation(context.Background(), env)
		return
	}

	// Fallback to legacy network broadcast
	_ = dc.network.BroadcastMessage(dc.networkID, typ.ProtocolMessage{Type: typ.MsgOperation, NetworkID: dc.networkID, SenderID: dc.network.GetPeerID(), Timestamp: time.Now().UnixMilli(), Payload: map[string]interface{}{"collection": dc.Name, "operation": op}})
}

func (dc *DistributedCollection) handleRemoteOperation(op typ.CRDTOperation) {
	ctx := context.Background()
	// Apply CRDT operation to local document
	existing, _ := dc.local.Find(ctx, op.DocumentID)
	var existingDist *typ.DistributedDocument
	if existing != nil {
		existingDist = resolver.ToDistributed(existing, op.PeerID)
	}

	result := resolver.ApplyOperation(existingDist, op)

	if result == nil {
		// delete
		_, _ = dc.local.Delete(ctx, op.DocumentID)
	} else if result.Deleted {
		_, _ = dc.local.Delete(ctx, op.DocumentID)
	} else {
		// upsert
		_, _ = dc.local.Insert(ctx, resolver.ToRegular(result))
	}

	// merge vector
	if dc.networkID != "" {
		syncState := dc.syncStates[dc.networkID]
		syncState.LocalVector = clock.Merge(syncState.LocalVector, op.Vector)
	}
}

func (dc *DistributedCollection) requestSync() error {
	if dc.networkID == "" {
		return errors.New("not attached to network")
	}
	syncState := dc.syncStates[dc.networkID]
	if syncState.SyncInProgress {
		return nil
	}
	syncState.SyncInProgress = true

	// Route sync request through p2pconsensus if available
	if dc.consensus != nil && dc.consensus.Enabled() {
		requester, ok := dc.consensus.(p2pconsensus.SyncRequester)
		if !ok {
			syncState.SyncInProgress = false
			return errors.New("consensus manager does not support sync")
		}
		req := p2pconsensus.SyncRequest{NetworkID: dc.networkID, Collection: dc.Name, VectorClock: map[string]int64(syncState.LocalVector)}
		go func() {
			err := requester.RequestSync(context.Background(), req)
			dc.mu.Lock()
			syncState.SyncInProgress = false
			dc.mu.Unlock()
			if err != nil {
				return
			}
		}()
		return nil
	}

	_ = dc.network.BroadcastMessage(dc.networkID, typ.ProtocolMessage{Type: typ.MsgSyncRequest, NetworkID: dc.networkID, SenderID: dc.network.GetPeerID(), Timestamp: time.Now().UnixMilli(), Payload: map[string]interface{}{"collection": dc.Name, "vector": syncState.LocalVector}})

	// Clear flag after timeout
	go func() {
		time.Sleep(10 * time.Second)
		syncState.SyncInProgress = false
	}()

	return nil
}

func (dc *DistributedCollection) handleSyncResponse(msg typ.ProtocolMessage) {
	payload, _ := msg.Payload.(map[string]interface{})
	opsIface, _ := payload["operations"].([]interface{})

	for _, oi := range opsIface {
		b, _ := jsonMarshal(oi)
		var op typ.CRDTOperation
		_ = jsonUnmarshal(b, &op)
		dc.handleRemoteOperation(op)
	}

	if dc.networkID != "" {
		syncState := dc.syncStates[dc.networkID]
		syncState.SyncInProgress = false
		syncState.LastSync = time.Now()
	}
}

func (dc *DistributedCollection) getCurrentVector() clock.VectorClock {
	if dc.networkID == "" {
		return clock.NewVectorClock()
	}
	if s, ok := dc.syncStates[dc.networkID]; ok {
		return clock.Clone(s.LocalVector)
	}
	return clock.NewVectorClock()
}

func (dc *DistributedCollection) pruneOperationLog() {
	if len(dc.operationLog) > dc.maxLogSize {
		dc.operationLog = dc.operationLog[len(dc.operationLog)-dc.maxLogSize:]
	}
}

// StreamBrackets returns a channel of live brackets from the underlying NRVStorage.
// Returns an error if the storage backend does not implement NRV streaming.
func (dc *DistributedCollection) StreamBrackets(ctx context.Context, goldOnly bool) (<-chan *nrv.Bracket, error) {
	nrvStore, ok := dc.local.store.(*stor.NRVStorage)
	if !ok {
		return nil, fmt.Errorf("collection %q: storage backend does not support NRV streaming", dc.Name)
	}
	return nrvStore.StreamBrackets(ctx, dc.Name, goldOnly)
}

// Helper JSON marshalling/unmarshalling helpers used by the example to avoid importing encoding/json repeatedly
func jsonMarshal(v interface{}) ([]byte, error)   { return json.Marshal(v) }
func jsonUnmarshal(b []byte, v interface{}) error { return json.Unmarshal(b, &v) }
